### ISSUE TITLE
Erroneous frame order in traceback for re-raised exceptions

### ISSUE THREAD

**Issue by ARF1** [micropython/micropython#2928]
For exceptions that are re-raised during exception handling (i.e. NOT chained), the frame ordering in a traceback appears to be erroneous:

```
MicroPython v1.8.7 on 2017-03-02; win32 version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>>
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== def some_fn():
===     foo()
===
=== def foo():
===     bar()
===
=== def bar():
===     raise RuntimeError('something')
===
=== def some_other_fn():
===     pass
===
=== try:
===     some_fn()
=== except Exception as e:
===     # do something
===     raise
=== finally:
===     some_other_fn()
===
===
Traceback (most recent call last):
  File "<stdin>", line 19, in <module>
  File "<stdin>", line 17, in <module>
  File "<stdin>", line 17, in <module>
  File "<stdin>", line 14, in <module>
  File "<stdin>", line 2, in some_fn
  File "<stdin>", line 5, in foo
  File "<stdin>", line 8, in bar
RuntimeError: something
>>>
```

There appear various things weird to me:

1. The order of the entries for lines 14-19 is reversed with respect to the order of execution.
2. The entry for line 17 is duplicated even though it is executed only once.

I believe the correct traceback should be:

```
Traceback (most recent call last):
  File "<stdin>", line 14, in <module>
  File "<stdin>", line 2, in some_fn
  File "<stdin>", line 5, in foo
  File "<stdin>", line 8, in bar
RuntimeError: something
```

If the frames for lines 17 & 19 are to be printed, it should be somehow clear that they were executed after the error was raised. The way it is now, the traceback is very misleading.

**Comment by ARF1** [comment]
Just for reference, on CPython the traceback is as follows:

```
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 2, in some_fn
  File "<stdin>", line 2, in foo
  File "<stdin>", line 2, in bar
RuntimeError: something
```

**Comment by dpgeorge** [comment]
The reason for the extra lines (at least line 19) is that a traceback is being added when the exception is re-raised after executing the finally clause.

**Comment by dpgeorge** [comment]
See #5032 for a fix.

**Comment by dpgeorge** [comment]
Fixed by 16f6169c88a5fd93b151149e561d177557fcc760 and 08c1fe556915c912598d5a5d5db0f5942f3a333d
